# Uso de la API de Spotify

In [26]:
import requests
import numpy as np
import pandas as pd

https://developer.spotify.com/documentation/web-api/reference/artists/

In [1]:
url_base = 'https://api.spotify.com/v1'

In [2]:
ep_artist = '/artists/{artist_id}'

Se debe conseguir el id del artista de donde se desea consumir los datos

In [3]:
id_thexx = '3iOvXCl6edW5Um0fXEBRXy'

In [4]:
url_base+ep_artist.format(artist_id = id_thexx)

'https://api.spotify.com/v1/artists/3iOvXCl6edW5Um0fXEBRXy'

In [6]:
r = requests.get(url_base+ep_artist.format(artist_id = id_thexx))

In [7]:
r.status_code

401

In [9]:
r.json()

{'error': {'status': 401, 'message': 'No token provided'}}

Como obtuvimos un status code 401 necesitamos dar de alta una aplicación en el servidor de spotify para obtener las credenciales

# Access token

In [10]:
token_url = 'https://accounts.spotify.com/api/token'

In [20]:
params = {'grant_type':'client_credentials'}

Se requiere tener nuestras credenciales tokenizadas con Base64: https://www.base64encode.org

In [21]:
header={'Authorization': 'Basic MTNiOWRiZGE0YjZmNDU4NThiZTBhZGRjMmRkYWU3MzU6ZjdiY2M0NjUzOTJlNDE1Yzg4ZGJhMTBjMGFkNmZmYzk='}

In [22]:
r = requests.post(token_url, data=params, headers=header)

In [23]:
r.status_code

200

In [25]:
r.json()

{'access_token': 'BQBhIU9bBat3nM0jyvwjL-56a6D8V97OziSta4463BoNIgkxgT_PgwPz9YWnKwuM84BCUjwVBiUXo-eUXuk',
 'token_type': 'Bearer',
 'expires_in': 3600,
 'scope': ''}

In [29]:
token = r.json()['access_token']

### Volvemos a correr lo que teníamos anteriormente pero ahora sí con el token 

In [30]:
header = {'Authorization': 'Bearer {}'.format(token)}

In [31]:
r = requests.get(url_base+ep_artist.format(artist_id = id_thexx), headers = header)

In [32]:
r.status_code

200

In [33]:
r.json()

{'external_urls': {'spotify': 'https://open.spotify.com/artist/3iOvXCl6edW5Um0fXEBRXy'},
 'followers': {'href': None, 'total': 3575877},
 'genres': ['downtempo', 'dream pop', 'indietronica'],
 'href': 'https://api.spotify.com/v1/artists/3iOvXCl6edW5Um0fXEBRXy',
 'id': '3iOvXCl6edW5Um0fXEBRXy',
 'images': [{'height': 640,
   'url': 'https://i.scdn.co/image/922163afa546eca372c27be4513a481de06e9481',
   'width': 640},
  {'height': 320,
   'url': 'https://i.scdn.co/image/8ef506656acfc2bc1449b9a963026bd7296df2e2',
   'width': 320},
  {'height': 160,
   'url': 'https://i.scdn.co/image/7e7c16105b54b9abcc056d25b3fb549372a8df38',
   'width': 160}],
 'name': 'The xx',
 'popularity': 70,
 'type': 'artist',
 'uri': 'spotify:artist:3iOvXCl6edW5Um0fXEBRXy'}

In [34]:
url_busqueda = 'https://api.spotify.com/v1/search'

In [35]:
search_params = {'q': "The+xx", 'type':'artist', 'market':'MX'}

In [38]:
busqueda = requests.get(url_busqueda, headers= header, params=search_params)

In [39]:
busqueda.status_code

200

In [40]:
busqueda.json()

{'artists': {'href': 'https://api.spotify.com/v1/search?query=The%2Bxx&type=artist&market=MX&offset=0&limit=20',
  'items': [{'external_urls': {'spotify': 'https://open.spotify.com/artist/3iOvXCl6edW5Um0fXEBRXy'},
    'followers': {'href': None, 'total': 3575877},
    'genres': ['downtempo', 'dream pop', 'indietronica'],
    'href': 'https://api.spotify.com/v1/artists/3iOvXCl6edW5Um0fXEBRXy',
    'id': '3iOvXCl6edW5Um0fXEBRXy',
    'images': [{'height': 640,
      'url': 'https://i.scdn.co/image/922163afa546eca372c27be4513a481de06e9481',
      'width': 640},
     {'height': 320,
      'url': 'https://i.scdn.co/image/8ef506656acfc2bc1449b9a963026bd7296df2e2',
      'width': 320},
     {'height': 160,
      'url': 'https://i.scdn.co/image/7e7c16105b54b9abcc056d25b3fb549372a8df38',
      'width': 160}],
    'name': 'The xx',
    'popularity': 70,
    'type': 'artist',
    'uri': 'spotify:artist:3iOvXCl6edW5Um0fXEBRXy'},
   {'external_urls': {'spotify': 'https://open.spotify.com/artist/4ah

In [41]:
import pandas as pd

In [42]:
df = pd.DataFrame(busqueda.json()['artists']['items'])

In [43]:
df

external_urls  \
0  {'spotify': 'https://open.spotify.com/artist/3...   
1  {'spotify': 'https://open.spotify.com/artist/4...   
2  {'spotify': 'https://open.spotify.com/artist/5...   
3  {'spotify': 'https://open.spotify.com/artist/5...   
4  {'spotify': 'https://open.spotify.com/artist/1...   
5  {'spotify': 'https://open.spotify.com/artist/4...   
6  {'spotify': 'https://open.spotify.com/artist/6...   
7  {'spotify': 'https://open.spotify.com/artist/4...   
8  {'spotify': 'https://open.spotify.com/artist/1...   
9  {'spotify': 'https://open.spotify.com/artist/2...   

                          followers                                genres  \
0  {'href': None, 'total': 3575877}  [downtempo, dream pop, indietronica]   
1      {'href': None, 'total': 278}                                    []   
2     {'href': None, 'total': 4848}                                    []   
3        {'href': None, 'total': 2}                                    []   
4     {'href': None, 'total': 1643}                                    []   
5       {'href': None, 'total': 51}                                    []   
6      {'href': None, 'total': 733}                                    []   
7        {'href': None, 'total': 8}                                    []   
8       {'href': None, 'total': 12}                                    []   
9      {'href': None, 'total': 104}                                    []   

                                                href                      id  \
0  https://api.spotify.com/v1/artists/3iOvXCl6edW...  3iOvXCl6edW5Um0fXEBRXy   
1  https://api.spotify.com/v1/artists/4ahQ9RPvFrb...  4ahQ9RPvFrbEV8KpZ5ZFEj   
2  https://api.spotify.com/v1/artists/5FtmpkzGoIs...  5FtmpkzGoIsZ0PfhPG4f1x   
3  https://api.spotify.com/v1/artists/5TZf05XNtrK...  5TZf05XNtrKFbA787u8pgn   
4  https://api.spotify.com/v1/artists/1rC1W1RD96S...  1rC1W1RD96Sp9nlds4YMOb   
5  https://api.spotify.com/v1/artists/4Hfk4E28xHD...  4Hfk4E28xHDci9vnoyDDxn   
6  https://api.spotify.com/v1/artists/6u9lHi4gk5e...  6u9lHi4gk5e4fq2UX1VgFO   
7  https://api.spotify.com/v1/artists/4RSmMzoRFs0...  4RSmMzoRFs0YMsvP3J31yA   
8  https://api.spotify.com/v1/artists/1fN9GfUqbyS...  1fN9GfUqbySx3sphqotvUe   
9  https://api.spotify.com/v1/artists/2p6ARPzuLzY...  2p6ARPzuLzYrrkXRKRRzc6   

                                              images  \
0  [{'height': 640, 'url': 'https://i.scdn.co/ima...   
1                                                 []   
2  [{'height': 640, 'url': 'https://i.scdn.co/ima...   
3  [{'height': 640, 'url': 'https://i.scdn.co/ima...   
4  [{'height': 640, 'url': 'https://i.scdn.co/ima...   
5                                                 []   
6  [{'height': 640, 'url': 'https://i.scdn.co/ima...   
7                                                 []   
8  [{'height': 640, 'url': 'https://i.scdn.co/ima...   
9  [{'height': 640, 'url': 'https://i.scdn.co/ima...   

                             name  popularity    type  \
0                          The xx          70  artist   
1                          The Xx           7  artist   
2                      THE XXXXXX          12  artist   
3                 The xxMirrorsxx          20  artist   
4                        The XX's           2  artist   
5                          The xx           0  artist   
6                         The XXX           1  artist   
7                       The XXVOT           0  artist   
8                    Red the Xxii           0  artist   
9  Dani Baraldés and The XXX Band           0  artist   

                                     uri  
0  spotify:artist:3iOvXCl6edW5Um0fXEBRXy  
1  spotify:artist:4ahQ9RPvFrbEV8KpZ5ZFEj  
2  spotify:artist:5FtmpkzGoIsZ0PfhPG4f1x  
3  spotify:artist:5TZf05XNtrKFbA787u8pgn  
4  spotify:artist:1rC1W1RD96Sp9nlds4YMOb  
5  spotify:artist:4Hfk4E28xHDci9vnoyDDxn  
6  spotify:artist:6u9lHi4gk5e4fq2UX1VgFO  
7  spotify:artist:4RSmMzoRFs0YMsvP3J31yA  
8  spotify:artist:1fN9GfUqbySx3sphqotvUe  
9  spotify:artist:

In [46]:
df.sort_values(by='popularity', ascending = False).iloc[0]['id']

'3iOvXCl6edW5Um0fXEBRXy'

In [47]:
import base64 

In [52]:
def get_token(client_id, client_secret):
    encoded = base64.b64encode(bytes(client_id+':'+client_secret, 'utf-8'))
    params = {'grant_type':'client_credentials'}
    header={'Authorization': 'Basic ' + str(encoded, 'utf-8')}
    r = requests.post('https://accounts.spotify.com/api/token', headers=header, data=params)
    if r.status_code != 200:
        print('Error en la request.', r.json())
        return None
    print('Token válido por {} segundos.'.format(r.json()['expires_in']))
    return r.json()['access_token']

In [53]:
client_id = '13b9dbda4b6f45858be0addc2ddae735'

In [54]:
client_secret = 'f7bcc465392e415c88dba10c0ad6ffc9'

In [55]:
token = get_token(client_id, client_secret)

Token válido por 3600 segundos.


In [56]:
header = {'Authorization': 'Bearer {}'.format(token)}

In [57]:
id_thexx

'3iOvXCl6edW5Um0fXEBRXy'

# Obtener información de un artista

In [62]:
ep_albums = '/artists/{artist_id}/albums'

In [63]:
url_base+ep_albums

'https://api.spotify.com/v1/artists/{artist_id}/albums'

In [64]:
params = {'country':'MX'}

In [68]:
albums_thexx = requests.get(url_base+ep_albums.format(artist_id=id_thexx), headers=header, params=params)

In [69]:
albums_thexx.status_code

200

In [72]:
albums_thexx.json()['items'][0]

{'album_group': 'album',
 'album_type': 'album',
 'artists': [{'external_urls': {'spotify': 'https://open.spotify.com/artist/3iOvXCl6edW5Um0fXEBRXy'},
   'href': 'https://api.spotify.com/v1/artists/3iOvXCl6edW5Um0fXEBRXy',
   'id': '3iOvXCl6edW5Um0fXEBRXy',
   'name': 'The xx',
   'type': 'artist',
   'uri': 'spotify:artist:3iOvXCl6edW5Um0fXEBRXy'}],
 'external_urls': {'spotify': 'https://open.spotify.com/album/2AKWZ0iZWyecVPCQT9Z11e'},
 'href': 'https://api.spotify.com/v1/albums/2AKWZ0iZWyecVPCQT9Z11e',
 'id': '2AKWZ0iZWyecVPCQT9Z11e',
 'images': [{'height': 640,
   'url': 'https://i.scdn.co/image/ab67616d0000b273425c6431e0b7170977c3a8cf',
   'width': 640},
  {'height': 300,
   'url': 'https://i.scdn.co/image/ab67616d00001e02425c6431e0b7170977c3a8cf',
   'width': 300},
  {'height': 64,
   'url': 'https://i.scdn.co/image/ab67616d00004851425c6431e0b7170977c3a8cf',
   'width': 64}],
 'name': 'I See You (Deluxe)',
 'release_date': '2017-06-23',
 'release_date_precision': 'day',
 'total_tr

#### Obtener lista de albumes (limit = 20)

In [80]:
lista_albums = [(album['id'],album['name']) for album in albums_thexx.json()['items']]
lista_albums

[('2AKWZ0iZWyecVPCQT9Z11e', 'I See You (Deluxe)'),
 ('6Zw6NKh3oIUhDRMOyBmsUU', 'I See You'),
 ('4SXTQ305LjMh2DBHA3CXIW', 'Coexist'),
 ('2av2ZSHlvD7rvLSsMvtYCG', 'xx'),
 ('7jIVGJg9wzec2svFKilP4w', 'Lips (Edu Imbernon Remix)'),
 ('1KwcjMbLdbmGBvHFRjNWCo', 'On Hold (Jamie xx Remix)'),
 ('3jw8KdtpBQGz3Cx4z7sLEm', 'A Violent Noise (Four Tet Remix)'),
 ('4tapXF4YEn2eGJxPyhb1QS', 'Hivern Remixes'),
 ('2jc0P2yriDQNz4Ns8oCv6a', 'Chained (LIAR Remix)'),
 ('77UpUTIMOspGHmpHwTqJtQ', 'Reunion (Edu Imbernon Remix)'),
 ('2mOBsYlzsidetXelSdcPDu', 'Fiction'),
 ('4bpEsVANclKtrZFj7KqWim', 'Fiction'),
 ('7fQHQnv7JCTa63MGAqqY5M', 'Innervisions Remixes'),
 ('3W6kQdUHO8P67ZNZSGRPuv', 'Jamie xx Edits'),
 ('6WbtzmiGvo97AhSe1Rpfb8', 'Sunset (Jamie Jones Remix)'),
 ('1k2I7tFdcul00H6o6jQeER', 'Sunset (Kim Ann Foxman Remix)'),
 ('3fyXjQKm3k6q1L2AiwtoqG', 'Angels (Four Tet Remix)'),
 ('3oPNd3hAWx8O4lQODh0mbV', 'Chained (John Talabot and Pional Blinded Remix)'),
 ('3Stu1ya9yxMOsowIBFovgr', 'Angels'),
 ('52Kxj5D44Vl9

In [81]:
album_ep = '/albums/{album_id}'
album_params = {'market':'MX'}

Extraer canciones de I See You (Deluxe)

In [83]:
ISYD_id = '2AKWZ0iZWyecVPCQT9Z11e'

In [84]:
ISYD = requests.get(url_base+album_ep.format(album_id=ISYD_id)+'/tracks', headers=header, params=params)
ISYD

<Response [200]>

In [86]:
ISYD.json()['items']

[{'artists': [{'external_urls': {'spotify': 'https://open.spotify.com/artist/3iOvXCl6edW5Um0fXEBRXy'},
    'href': 'https://api.spotify.com/v1/artists/3iOvXCl6edW5Um0fXEBRXy',
    'id': '3iOvXCl6edW5Um0fXEBRXy',
    'name': 'The xx',
    'type': 'artist',
    'uri': 'spotify:artist:3iOvXCl6edW5Um0fXEBRXy'}],
  'available_markets': ['AD',
   'AE',
   'AL',
   'AR',
   'AT',
   'AU',
   'BA',
   'BE',
   'BG',
   'BH',
   'BO',
   'BR',
   'BY',
   'CA',
   'CH',
   'CL',
   'CO',
   'CR',
   'CY',
   'CZ',
   'DE',
   'DK',
   'DO',
   'DZ',
   'EC',
   'EE',
   'EG',
   'ES',
   'FI',
   'FR',
   'GB',
   'GR',
   'GT',
   'HK',
   'HN',
   'HR',
   'HU',
   'ID',
   'IE',
   'IL',
   'IN',
   'IS',
   'IT',
   'JO',
   'JP',
   'KW',
   'KZ',
   'LB',
   'LI',
   'LT',
   'LU',
   'LV',
   'MA',
   'MC',
   'MD',
   'ME',
   'MK',
   'MT',
   'MX',
   'MY',
   'NI',
   'NL',
   'NO',
   'NZ',
   'OM',
   'PA',
   'PE',
   'PH',
   'PL',
   'PS',
   'PT',
   'PY',
   'QA',
   'RO',
   

Obtener canciones del album

In [87]:
[(track['id'], track['name']) for track in ISYD.json()['items']]

[('2CRi1nkSTfPCy6BbzRagO3', 'Dangerous'),
 ('0qPHDXv7tpY4oDs8M8xe2E', 'Say Something Loving'),
 ('4QIvsXT9XIvU23GDQccFw0', 'Lips'),
 ('0nXucRZlb2VMOVmqo2exYT', 'A Violent Noise'),
 ('0SSuzo9xyao2rvLLPr6CUI', 'Performance'),
 ('1aRMQBLkBEGBn63QxK6mOL', 'Replica'),
 ('1et1bq5DmAQC7OtGRJk1DG', 'Brave For You'),
 ('7fEVEn2QxhSCLewdFtPnhX', 'On Hold'),
 ('4MkJRmOUbcnm5dOUVE29pL', 'I Dare You'),
 ('4E2DAePBxomdzuKzITMtmT', 'Test Me'),
 ('10QOoEpAFpA45zJqp8culW', 'Naive'),
 ('51fK4tBe0X09zb6KgGSsKX', 'Seasons Run'),
 ('6shypNQ3DRHzSyCMGt3TPk', 'Brave For You - Marfa Demo')]

# Función para obtener discografía

In [88]:
def obtener_discografia(artist_id, token, return_name = False, page_limit =50, country = None):
    url = f'https://api.spotify.com/v1/artists/{artist_id}/albums'
    header = {'Authorization': f'Bearer {token}'}
    params = {'limit': page_limit, 
              'offset': 0,
              'country': country}
    lista = []
    r = requests.get(url, params=params, headers=header)
    
    if r.status_code != 200:
        print('Error en request.', r.json())
        return None
    
    if return_name:
        lista += [(item['id'], item['name']) for item in r.json()['items']]
    else:
        lista += [item['id'] for item in r.json()['items']]
        
    while r.json()['next']:
        r = requests.get(r.json()['next'], headers=header) # El resto de los parámetros están dentro de la URL
        if return_name:
            lista += [(item['id'], item['name']) for item in r.json()['items']]
        else:
            lista += [item['id'] for item in r.json()['items']]
    
    return lista

In [89]:
obtener_discografia(id_thexx, token, return_name=True)

[('2AKWZ0iZWyecVPCQT9Z11e', 'I See You (Deluxe)'),
 ('6Zw6NKh3oIUhDRMOyBmsUU', 'I See You'),
 ('4SXTQ305LjMh2DBHA3CXIW', 'Coexist'),
 ('6aAMm8VNTcio4a2OrveBtP', 'Coexist'),
 ('6tzQKMilI02kn1lzLklDI8', 'xx'),
 ('2av2ZSHlvD7rvLSsMvtYCG', 'xx'),
 ('7jIVGJg9wzec2svFKilP4w', 'Lips (Edu Imbernon Remix)'),
 ('1KwcjMbLdbmGBvHFRjNWCo', 'On Hold (Jamie xx Remix)'),
 ('3jw8KdtpBQGz3Cx4z7sLEm', 'A Violent Noise (Four Tet Remix)'),
 ('4tapXF4YEn2eGJxPyhb1QS', 'Hivern Remixes'),
 ('2jc0P2yriDQNz4Ns8oCv6a', 'Chained (LIAR Remix)'),
 ('77UpUTIMOspGHmpHwTqJtQ', 'Reunion (Edu Imbernon Remix)'),
 ('2mOBsYlzsidetXelSdcPDu', 'Fiction'),
 ('4bpEsVANclKtrZFj7KqWim', 'Fiction'),
 ('7fQHQnv7JCTa63MGAqqY5M', 'Innervisions Remixes'),
 ('3W6kQdUHO8P67ZNZSGRPuv', 'Jamie xx Edits'),
 ('6WbtzmiGvo97AhSe1Rpfb8', 'Sunset (Jamie Jones Remix)'),
 ('1k2I7tFdcul00H6o6jQeER', 'Sunset (Kim Ann Foxman Remix)'),
 ('3fyXjQKm3k6q1L2AiwtoqG', 'Angels (Four Tet Remix)'),
 ('3oPNd3hAWx8O4lQODh0mbV', 'Chained (John Talabot and Pion

# Función para obtener los tracks del album

In [90]:
def obtener_tracks(album_id, token, return_name=False, page_limit=50, country=None):
    url=f'https://api.spotify.com/v1/albums/{album_id}/tracks'
    header = {'Authorization': f'Bearer {token}'}
    params = {'limit': page_limit, 
              'offset': 0,
              'market': country}
    lista = []
    r = requests.get(url, params=params, headers=header)
    
    if r.status_code != 200:
        print('Error en request.', r.json())
        return None
    
    if return_name:
        lista += [(item['id'], item['name']) for item in r.json()['items']]
    else:
        lista += [item['id'] for item in r.json()['items']]
        
    while r.json()['next']:
        r = requests.get(r.json()['next'], headers=header) # El resto de los parámetros están dentro de la URL
        if return_name:
            lista += [(item['id'], item['name']) for item in r.json()['items']]
        else:
            lista += [item['id'] for item in r.json()['items']]
    
    return lista

In [91]:
obtener_tracks('1hS6yQqEmHAsZeDxu0wtWK', token, return_name=True, country='MX')

[('3NXLWhPkRRy8EBUUmgfoWi', 'Islands - Untold Remix'),
 ('2xCb4ceu74G1VxhSCgC4ba', 'Islands - The Blue Nile Remix'),
 ('4MqalvDBsY4wfSokuOwUvU', 'Islands - Nosaj Thing Remix'),
 ('1NxzYu3A2fs2aID0wuLR8Z', 'Islands - Delorean Remix'),
 ('74a2hGmR7Xc79ryOXdn5Bb', 'Islands - Falty DL Remix')]

Ahora podemos usar ambas funciones en combinación para obtener la lista de temas disponibles en México de The XX, cada una con referencia a su album

In [92]:
for album in obtener_discografia(id_thexx, token, return_name=True, country='MX'):
    print(album[1])
    for track in obtener_tracks(album[0], token, return_name=True, country='MX'):
        print('\t',track[1])   

I See You (Deluxe)
	 Dangerous
	 Say Something Loving
	 Lips
	 A Violent Noise
	 Performance
	 Replica
	 Brave For You
	 On Hold
	 I Dare You
	 Test Me
	 Naive
	 Seasons Run
	 Brave For You - Marfa Demo
I See You
	 Dangerous
	 Say Something Loving
	 Lips
	 A Violent Noise
	 Performance
	 Replica
	 Brave For You
	 On Hold
	 I Dare You
	 Test Me
Coexist
	 Angels
	 Chained
	 Fiction
	 Try
	 Reunion
	 Sunset
	 Missing
	 Tides
	 Unfold
	 Swept Away
	 Our Song
xx
	 Intro
	 VCR
	 Crystalised
	 Islands
	 Heart Skipped A Beat
	 Fantasy
	 Shelter
	 Basic Space
	 Infinity
	 Night Time
	 Stars
Lips (Edu Imbernon Remix)
	 Lips - Edu Imbernon Remix
On Hold (Jamie xx Remix)
	 On Hold - Jamie xx Remix
A Violent Noise (Four Tet Remix)
	 A Violent Noise - Four Tet Remix
Hivern Remixes
	 Fiction - Marc Piñol Remix
	 Reunion - Mistakes Are Ok Remix
	 Missing - Round Remix
	 Swept Away - New Jackson Remix
Chained (LIAR Remix)
	 Chained - LIAR Remix
Reunion (Edu Imbernon Remix)
	 Reunion - Edu Imbernon Remi

# Vista previa de una canción

In [93]:
preview_url = 'https://p.scdn.co/mp3-preview/5a8286d973fc9f6dfb0a6fe44fd1c4ebf32a0f40?cid=13b9dbda4b6f45858be0addc2ddae735'

In [95]:
preview = requests.get(preview_url)
preview

<Response [200]>

In [96]:
import IPython.display as ipd

In [98]:
preview.content

b'ID3\x04\x00\x00\x00\x00\x10\x00GEOB\x00\x00\x0cm\x00\x00\x00application/json\x00metadata.json\x00None\x00{"loop_duration": 16.006569999999996, "start_time": 81.89901707913117, "downbeats": [0.1250229208688296, 2.1233729208688317, 4.123972920868823, 6.128852920868823, 8.129522920868837, 10.127902920868834, 12.125222920868836, 14.129842920868825, 16.131592920868826, 18.130632920868834, 20.125612920868832, 22.12283292086883, 24.118542920868833, 26.119762920868837, 28.121152920868823], "beats": [0.1250229208688296, 0.6241229208688281, 1.1235929208688304, 1.623872920868834, 2.1233729208688317, 2.622772920868826, 3.123162920868836, 3.623632920868829, 4.123972920868823, 4.625302920868833, 5.127182920868833, 5.628732920868828, 6.128852920868823, 6.628452920868824, 7.1288729208688295, 7.629892920868826, 8.129522920868837, 8.628212920868833, 9.12820292086883, 9.628832920868831, 10.127902920868834, 10.625632920868824, 11.125162920868831, 11.625842920868834, 12.125222920868836, 12.62417292086883

In [97]:
ipd.Audio(preview.content)